### PR TITLE
Store original roles with user login state.

### PR DESCRIPTION
**Pull request by mdwn**
The user's original static roles are now stored along with the user login state. This is necessary to properly recalculate user access on access list membership change, as SSO users will expire after 24 hours, making it impossible to refer to the original user definition to determine the new role membership for the user.

### CODE REVIEW

**Comment by zmb3** [comment]
Any concerns with backwards compatibility here? (Doesn't look like it, though a subsequent PR probably needs to be careful to handle the case where one of the original roles no longer exists..)

**Comment by mdwn** [comment]
> Any concerns with backwards compatibility here? (Doesn't look like it, though a subsequent PR probably needs to be careful to handle the case where one of the original roles no longer exists..)

No, but totally agreed. I'll need to ensure that the consumers of this look after this situation carefully.

File: api/proto/teleport/userloginstate/v1/userloginstate.proto [hunk 1]
```diff
@@ -40,4 +40,7 @@ message Spec {
 
   // user_type is the type of user this state represents.
   string user_type = 3;
+
+  // original_roles are the user roles that the user originally had prior to the user login state being generated.
```
Comment (zmb3) [inline api/proto/teleport/userloginstate/v1/userloginstate.proto:44]: When is a user login state generated? Is this prior to getting access from an access list?

File: lib/auth/userloginstate/generator_test.go [hunk 2]
```diff
@@ -257,14 +228,9 @@ func TestAccessLists(t *testing.T) {
...
-					"role1",
-				},
-				trait.Traits{
-					"trait1": []string{"value1", "value2"},
-					"trait2": []string{"value3", "value4"},
-					"trait3": []string{"value5", "value6"},
-				}),
+				[]string{},
+				[]string{"role1"},
+				trait.Traits{"trait1": {"value1", "value2"}, "trait2": {"value3", "value4"}, "trait3": {"value5", "value6"}}),
```
Comment (zmb3) [inline lib/auth/userloginstate/generator_test.go:233]: Break this long line up? (like you had it before)

File: lib/auth/userloginstate/generator_test.go [hunk 3]
```diff
@@ -257,14 +228,9 @@ func TestAccessLists(t *testing.T) {
...
 			expected: newUserLoginState(t, "user",
-				[]string{
-					"role1",
-				},
-				trait.Traits{
-					"trait1": []string{"value1", "value2"},
-					"trait2": []string{"value3", "value4"},
-					"trait3": []string{"value5", "value6"},
-				}),
+				[]string{},
```
Comment (zmb3) [inline lib/auth/userloginstate/generator_test.go:231]: Prefer `nil` to an empty slice, unless `newUserLoginState` actually cares about the difference.

**Review by zmb3 — APPROVED** [review]

File: api/proto/teleport/userloginstate/v1/userloginstate.proto (hunk 1 above)
Comment (tigrato) [reply api/proto/teleport/userloginstate/v1/userloginstate.proto:44]: I agree with Zac. Some more details here would be helpful

**Review by tigrato — APPROVED** [review]

**Pull request merged**